### PR TITLE
feat: Expose unpacked topk weights for routed moe (fp4)

### PR DESCRIPTION
**Pull request by aleozlx**
## 📌 Description

Summary

  Add support for pre-computed routing with unpacked format in trtllm_fp4_block_scale_routed_moe, and
  improve code clarity with explicit routing mode enum.

  Changes

  New Feature: Unpacked Pre-computed Routing (Mode 3)

  trtllm_fp4_block_scale_routed_moe now accepts routing input in two formats:

  # Packed format (existing behavior)
  trtllm_fp4_block_scale_routed_moe(packed_tensor, ...)

  # Unpacked format (new - pass tuple)
  trtllm_fp4_block_scale_routed_moe((topk_ids, topk_weights), ...)

  This is a backwards-compatible enhancement - existing code continues to work unchanged.

  Code Clarity Improvements

  - Added RoutingInputMode enum in both C++ and Python to explicitly define three routing modes:
    - FromLogits: Compute routing from logits (Mode 1)
    - PackedPrecomputed: Pre-computed with packed (score << 16 | id) format (Mode 2)
    - UnpackedPrecomputed: Pre-computed with separate tensors (Mode 3)
  - Renamed expert_weights → topk_weights in public APIs for consistency (MLP weights like
  fc1_expert_weights unchanged)
  - Added routing_input_mode as first parameter to internal C++ function for explicit mode selection

  Files Modified

  - flashinfer/fused_moe/core.py - Python API with RoutingInputMode enum and tuple support
  - csrc/trtllm_fused_moe_kernel_launcher.cu - C++ launcher with RoutingInputMode enum
  - tests/moe/test_trtllm_gen_routed_fused_moe.py - Extended tests for both formats
  - INTEGRATION_STATUS.md - Documentation (to be removed after merge)

  Test plan

  - test_trtllm_gen_fused_moe.py - All passed (Mode 1: FromLogits)
  - test_trtllm_gen_routed_fused_moe.py with routing_format=packed - All passed (Mode 2)
  - test_trtllm_gen_routed_fused_moe.py with routing_format=unpacked - All passed (Mode 3)

  pytest tests/moe/test_trtllm_gen_routed_fused_moe.py -v

  🤖 Generated with https://claude.ai/code (fully reviewed)

## 🔍 Related Issues

#2373

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

```
$ pytest tests/moe/test_trtllm_gen_routed_fused_moe.py
3456 passed in 336.66s (0:05:36)
```

## Summary by CodeRabbit

* **New Features**
  * Added a RoutingInputMode with three modes and end-to-end support for packed or unpacked precomputed routing inputs; intermediate routed outputs now use topk_weights.

* **Public API**
  * Core APIs now accept precomputed routing IDs/weights and an explicit routing mode to enable mode-aware routing paths.

* **Tests**
  * Parameterized tests added for "packed" and "unpacked" routing formats.

* **Documentation**
  * Docstrings and return descriptions updated to reflect new routing formats and topk_weights semantics.

### CODE REVIEW

**Comment by aleozlx** [comment]
@claude hello

**Comment by aleozlx** [comment]
@claude can you make a similar PR but for fp8 routed_moe?

**Comment by aleozlx** [comment]
That last fix (dafc67d) passed all tests

```
 pytest tests/moe/test_trtllm_gen_routed_fused_moe.py -x
```

```
 3488 passed in 363.64s (0:06:03)
```

**Comment by aleozlx** [comment]
@flashinfer-bot run

**Comment by aleozlx** [comment]
posting latest test result again for good measure

```
$ git log -1
commit a06a0fd23a4b469fc838f23fd1ac084f3f107402 (HEAD -> topk_weights, origin/topk_weights)
Merge: dafc67d9 1d350ae5
Author: Alex Yang <aleyang@nvidia.com>
Date:   Wed Feb 11 18:25:20 2026 -0800

    Merge branch 'main' into topk_weights

$ pytest tests/moe/test_trtllm_gen_routed_fused_moe.py
3488 passed in 954.71s (0:15:54)
```

**Comment by aleozlx** [comment]
resolved conflicts again, removing `ready` label pending new test results

**Comment by aleozlx** [comment]
@IwakuraRein could you give an approval for code review? (as i can't give approval to my own PR)

then i'll tag others who have permission to merge after new testing

decoupling and trying to unstick this PR that's been here for a while...

**Review by samuellees — APPROVED** [review]
LGTM, CI passed
